testing/test_junitxml.py

```diff
@@ -4,51 +4,60 @@
 from pathlib import Path
 from typing import cast
 from typing import List
+from typing import Optional
 from typing import Tuple
 from typing import TYPE_CHECKING
+from typing import TypeVar
+from typing import Union
 from xml.dom import minidom
 
-import py
 import xmlschema
 
 import pytest
 from _pytest.config import Config
 from _pytest.junitxml import bin_xml_escape
 from _pytest.junitxml import LogXML
+from _pytest.monkeypatch import MonkeyPatch
+from _pytest.pytester import Pytester
+from _pytest.pytester import RunResult
 from _pytest.reports import BaseReport
 from _pytest.reports import TestReport
 from _pytest.store import Store
 
+T = TypeVar("T")
+
 
 @pytest.fixture(scope="session")
-def schema():
+def schema() -> xmlschema.XMLSchema:
     """Return an xmlschema.XMLSchema object for the junit-10.xsd file."""
     fn = Path(__file__).parent / "example_scripts/junit-10.xsd"
     with fn.open() as f:
         return xmlschema.XMLSchema(f)
 
 
 @pytest.fixture
-def run_and_parse(testdir, schema):
+def run_and_parse(pytester: Pytester, schema: xmlschema.XMLSchema) -> T:
```

Comment:
The `T` here I think is not what we want. In general a generic in return position which is not bound by any inputs is probably not useful, because it does not provide any type safety, the caller can define it to be whatever it wants.

I think `typing.Callable` is not able to type callables with args and kwargs so it'd be hard to type strictly.  [Callable protocol](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols) can do it but they require newer Python version than pytest supports. So I suggest just using a callable class like this:

```py
class RunAndParse:
    def __init__(self, pytester: Pytester, schema: xmlschema.XmlSchema) -> None:
        self.pytester = pytester
        self.schema = schema

    def __call__(
        self, *args: str, family: Optional[str] = "xunit1"
    ) -> Tuple[RunResult, "DomNode"]:
        if family:
            args = ("-o", "junit_family=" + family) + args
        xml_path = self.pytester.path.joinpath("junit.xml")
        result = self.pytester.runpytest("--junitxml=%s" % xml_path, *args)
        if family == "xunit2":
            with xml_path.open() as f:
                self.schema.validate(f)
        xmldoc = minidom.parse(str(xml_path))
        return result, DomNode(xmldoc)


@pytest.fixture
def run_and_parse(pytester: Pytester, schema: xmlschema.XmlSchema) -> RunAndParse:
    """Fixture that returns a function that can be used to execute pytest and
    return the parsed ``DomNode`` of the root xml node.

    The ``family`` parameter is used to configure the ``junit_family`` of the written report.
    "xunit2" is also automatically validated against the schema.
    """
    return RunAndParse(pytester, schema)
```

Then you can also type the `run_and_parse` fixture arguments in tests as `run_and_parse: RunAndParse`.